From patchwork Tue May  4 11:05:20 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: "Alejandro Colomar \\(man-pages\\)"
 <alx.manpages@gmail.com>
X-Patchwork-Id: 43252
Return-Path: <libc-alpha-bounces@sourceware.org>
X-Original-To: patchwork@sourceware.org
Delivered-To: patchwork@sourceware.org
Received: from server2.sourceware.org (localhost [IPv6:::1])
	by sourceware.org (Postfix) with ESMTP id 2C1913982070;
	Tue,  4 May 2021 11:05:49 +0000 (GMT)
DKIM-Filter: OpenDKIM Filter v2.11.0 sourceware.org 2C1913982070
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=sourceware.org;
	s=default; t=1620126349;
	bh=jBexOv/slXZrUqoBk/6dME5LyTQkC5h8UHzK8vFSGV0=;
	h=To:Subject:Date:In-Reply-To:References:List-Id:List-Unsubscribe:
	 List-Archive:List-Post:List-Help:List-Subscribe:From:Reply-To:Cc:
	 From;
	b=qatNBway8DiTlROQUBxHhia0YfTpPjuco+U0buI2iYy7eqfSkRZt2Yygz724Jcft5
	 m9zCsCM1iqkrrRX6Mp90tn5n2aMMllXLMR6tl9dT6HF63RWWjUuNcQhiYMlbVBCd1h
	 lXziCfaijKOSXtWGTk2EfM7/VNXVrKh9iZKN5GwE=
X-Original-To: libc-alpha@sourceware.org
Delivered-To: libc-alpha@sourceware.org
Received: from mail-wr1-x42c.google.com (mail-wr1-x42c.google.com
 [IPv6:2a00:1450:4864:20::42c])
 by sourceware.org (Postfix) with ESMTPS id 319813950C61
 for <libc-alpha@sourceware.org>; Tue,  4 May 2021 11:05:45 +0000 (GMT)
DMARC-Filter: OpenDMARC Filter v1.3.2 sourceware.org 319813950C61
Received: by mail-wr1-x42c.google.com with SMTP id h4so8882079wrt.12
 for <libc-alpha@sourceware.org>; Tue, 04 May 2021 04:05:45 -0700 (PDT)
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=1e100.net; s=20161025;
 h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
 :references:mime-version:content-transfer-encoding;
 bh=jBexOv/slXZrUqoBk/6dME5LyTQkC5h8UHzK8vFSGV0=;
 b=XigS85RQMLbT6IYLNgszsX2L/diEK/Kao2jaMGC7diDsAUr6BacyVMNdpggkpG3E7t
 135KHGY3mfpTS+fiE7DqL2MLfvqHBlNkLTpz1Ni+xITtEYr7bUVszIP4eV4pv2z1mGOp
 ciBnEw7j0Zc1gbnoMq5Qsh3TYMroKYrz2ISL0yhwN0n+o/beoPo5PapmKPB8wnrVdBuP
 6o2Zl8gzl4OJ6UYr0zdTKSejpOlGm9OMH9RXLYxPy1j0h6wpGmGX7WUSvPT4U625MZ+u
 N+XwHtqumZcdqQowC87HOpdTB53Kso0+f3Umq4VweglLK6XcIEvXmxHtjJnKVjcoFCeb
 RzVg==
X-Gm-Message-State: AOAM531ES0V5L1LLxNmtT7MK6x2nggBh5+3WdVazxHmrra6nVtzI91NK
 6QO5ZsGwRwOJJzJQUyHVvdU=
X-Google-Smtp-Source: 
 ABdhPJw47QkoJpJSP1YPM0MC+fPhs/05hHrWzAdxeUw5qEExJJXq+uqU1z5Qjad2cZ6BA+EyTVcMTw==
X-Received: by 2002:adf:e811:: with SMTP id o17mr14669020wrm.71.1620126344302;
 Tue, 04 May 2021 04:05:44 -0700 (PDT)
Received: from sqli.sqli.com ([195.53.121.100])
 by smtp.googlemail.com with ESMTPSA id
 o17sm16769939wrs.48.2021.05.04.04.05.42
 (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
 Tue, 04 May 2021 04:05:43 -0700 (PDT)
To: mtk.manpages@gmail.com,
	linux-man@vger.kernel.org
Subject: [RFC v2] bpf.2: Use standard types and attributes
Date: Tue,  4 May 2021 13:05:20 +0200
Message-Id: <20210504110519.16097-1-alx.manpages@gmail.com>
X-Mailer: git-send-email 2.31.1
In-Reply-To: <20210423230609.13519-1-alx.manpages@gmail.com>
References: <20210423230609.13519-1-alx.manpages@gmail.com>
MIME-Version: 1.0
X-Spam-Status: No, score=-12.3 required=5.0 tests=BAYES_00, DKIM_SIGNED,
 DKIM_VALID, DKIM_VALID_AU, DKIM_VALID_EF, FREEMAIL_FROM, GIT_PATCH_0,
 RCVD_IN_DNSWL_NONE, SPF_HELO_NONE, SPF_PASS,
 TXREP autolearn=ham autolearn_force=no version=3.4.2
X-Spam-Checker-Version: SpamAssassin 3.4.2 (2018-09-13) on
 server2.sourceware.org
X-BeenThere: libc-alpha@sourceware.org
X-Mailman-Version: 2.1.29
Precedence: list
List-Id: Libc-alpha mailing list <libc-alpha.sourceware.org>
List-Unsubscribe: <https://sourceware.org/mailman/options/libc-alpha>,
 <mailto:libc-alpha-request@sourceware.org?subject=unsubscribe>
List-Archive: <https://sourceware.org/pipermail/libc-alpha/>
List-Post: <mailto:libc-alpha@sourceware.org>
List-Help: <mailto:libc-alpha-request@sourceware.org?subject=help>
List-Subscribe: <https://sourceware.org/mailman/listinfo/libc-alpha>,
 <mailto:libc-alpha-request@sourceware.org?subject=subscribe>
X-Patchwork-Original-From: Alejandro Colomar via Libc-alpha
 <libc-alpha@sourceware.org>
From: "Alejandro Colomar \\(man-pages\\)" <alx.manpages@gmail.com>
Reply-To: Alejandro Colomar <alx.manpages@gmail.com>
Cc: Alejandro Colomar <alx.manpages@gmail.com>,
 glibc <libc-alpha@sourceware.org>, LKML <linux-kernel@vger.kernel.org>,
 David Laight <David.Laight@ACULAB.COM>, GCC <gcc-patches@gcc.gnu.org>,
 bpf <bpf@vger.kernel.org>, Alexei Starovoitov <alexei.starovoitov@gmail.com>,
 Joseph Myers <joseph@codesourcery.com>
Errors-To: libc-alpha-bounces@sourceware.org
Sender: "Libc-alpha" <libc-alpha-bounces@sourceware.org>

Some manual pages are already using C99 syntax for integral
types 'uint32_t', but some aren't.  There are some using kernel
syntax '__u32'.  Fix those.

Some pages also document attributes, using GNU syntax
'__attribute__((xxx))'.  Update those to use the shorter and more
portable C11 keywords such as 'alignas()' when possible, and C2x
syntax '[[gnu::xxx]]' elsewhere, which hasn't been standardized
yet, but is already implemented in GCC, and available through
either --std=c2x or any of the --std=gnu... options.

The standard isn't very clear on how to use alignas() or
[[]]-style attributes, so the following link is useful in the case
of 'alignas()' and '[[gnu::aligned()]]':
<https://stackoverflow.com/q/67271825/6872717>

Signed-off-by: Alejandro Colomar <alx.manpages@gmail.com>
Cc: LKML <linux-kernel@vger.kernel.org>
Cc: glibc <libc-alpha@sourceware.org>
Cc: GCC <gcc-patches@gcc.gnu.org>
Cc: Alexei Starovoitov <alexei.starovoitov@gmail.com>
Cc: bpf <bpf@vger.kernel.org>
Cc: David Laight <David.Laight@ACULAB.COM>
Cc: Zack Weinberg <zackw@panix.com>
Cc: Joseph Myers <joseph@codesourcery.com>
Nacked-by: Alexei Starovoitov <ast@kernel.org>
Nacked-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---
 man2/bpf.2 | 49 ++++++++++++++++++++++++-------------------------
 1 file changed, 24 insertions(+), 25 deletions(-)

diff --git a/man2/bpf.2 b/man2/bpf.2
index 6e1ffa198..04b8fbcef 100644
--- a/man2/bpf.2
+++ b/man2/bpf.2
@@ -186,41 +186,40 @@ commands:
 .PP
 .in +4n
 .EX
-union bpf_attr {
+union [[gnu::aligned(8)]] bpf_attr {
     struct {    /* Used by BPF_MAP_CREATE */
-        __u32         map_type;
-        __u32         key_size;    /* size of key in bytes */
-        __u32         value_size;  /* size of value in bytes */
-        __u32         max_entries; /* maximum number of entries
-                                      in a map */
+        uint32_t    map_type;
+        uint32_t    key_size;    /* size of key in bytes */
+        uint32_t    value_size;  /* size of value in bytes */
+        uint32_t    max_entries; /* maximum number of entries
+                                    in a map */
     };
 
-    struct {    /* Used by BPF_MAP_*_ELEM and BPF_MAP_GET_NEXT_KEY
-                   commands */
-        __u32         map_fd;
-        __aligned_u64 key;
+    struct {    /* Used by BPF_MAP_*_ELEM and BPF_MAP_GET_NEXT_KEY commands */
+        uint32_t            map_fd;
+        uint64_t alignas(8) key;
         union {
-            __aligned_u64 value;
-            __aligned_u64 next_key;
+            uint64_t alignas(8) value;
+            uint64_t alignas(8) next_key;
         };
-        __u64         flags;
+        uint64_t            flags;
     };
 
     struct {    /* Used by BPF_PROG_LOAD */
-        __u32         prog_type;
-        __u32         insn_cnt;
-        __aligned_u64 insns;      /* \(aqconst struct bpf_insn *\(aq */
-        __aligned_u64 license;    /* \(aqconst char *\(aq */
-        __u32         log_level;  /* verbosity level of verifier */
-        __u32         log_size;   /* size of user buffer */
-        __aligned_u64 log_buf;    /* user supplied \(aqchar *\(aq
-                                     buffer */
-        __u32         kern_version;
-                                  /* checked when prog_type=kprobe
-                                     (since Linux 4.1) */
+        uint32_t            prog_type;
+        uint32_t            insn_cnt;
+        uint64_t alignas(8) insns;     /* \(aqconst struct bpf_insn *\(aq */
+        uint64_t alignas(8) license;   /* \(aqconst char *\(aq */
+        uint32_t            log_level; /* verbosity level of verifier */
+        uint32_t            log_size;  /* size of user buffer */
+        uint64_t alignas(8) log_buf;   /* user supplied \(aqchar *\(aq
+                                          buffer */
+        uint32_t            kern_version;
+                                       /* checked when prog_type=kprobe
+                                          (since Linux 4.1) */
 .\"                 commit 2541517c32be2531e0da59dfd7efc1ce844644f5
     };
-} __attribute__((aligned(8)));
+};
 .EE
 .in
 .\"
